akka-stream-tests/src/test/scala/akka/stream/remote/StreamRefsSpec.scala

```diff
@@ -0,0 +1,290 @@
+
+  override protected def beforeTermination(): Unit =
+    TestKit.shutdownActorSystem(remoteSystem)
+
+  val p = TestProbe()
+
+  // obtain the remoteActor ref via selection in order to use _real_ remoting in this test
+  val remoteActor = {
+    val it = remoteSystem.actorOf(DatasourceActor.props(p.ref), "remoteActor")
+    val remoteAddress = remoteSystem.asInstanceOf[ActorSystemImpl].provider.getDefaultAddress
+    system.actorSelection(it.path.toStringWithAddress(remoteAddress)) ! Identify("hi")
+    expectMsgType[ActorIdentity].ref.get
+  }
+
+  "A SourceRef" must {
+
+    "send messages via remoting" in {
+      remoteActor ! "give"
+      val sourceRef = expectMsgType[SourceRef[String]]
+
+      sourceRef
+        .runWith(Sink.actorRef(p.ref, "<COMPLETE>"))
+
+      p.expectMsg("hello")
+      p.expectMsg("world")
+      p.expectMsg("<COMPLETE>")
+    }
+
+    "fail when remote source failed" in {
+      remoteActor ! "give-fail"
+      val sourceRef = expectMsgType[SourceRef[String]]
+
+      sourceRef
+        .runWith(Sink.actorRef(p.ref, "<COMPLETE>"))
+
+      val f = p.expectMsgType[Failure]
+      f.cause.getMessage should include("Remote stream (")
+      // actor name here, for easier identification
+      f.cause.getMessage should include("failed, reason: Booooom!")
+    }
+
+    "complete properly when remote source is empty" in {
+      // this is a special case since it makes sure that the remote stage is still there when we connect to it
+
+      remoteActor ! "give-complete-asap"
+      val sourceRef = expectMsgType[SourceRef[String]]
+
+      sourceRef
+        .runWith(Sink.actorRef(p.ref, "<COMPLETE>"))
+
+      p.expectMsg("<COMPLETE>")
+    }
+
+  }
+
+  "A SinkRef" must {
+
+    "xoxo receive elements via remoting" in {
+
+      remoteActor ! "receive"
+      val remoteSink: SinkRef[String] = expectMsgType[SinkRef[String]]
+
+      Source("hello" :: "world" :: Nil)
+        .to(remoteSink)
+        .run()
+
+      p.expectMsg("hello")
+      p.expectMsg("world")
+      p.expectMsg("<COMPLETE>")
+    }
+
+    "fail origin if remote Sink gets a failure" in {
+
+      remoteActor ! "receive"
+      val remoteSink: SinkRef[String] = expectMsgType[SinkRef[String]]
+
+      val remoteFailureMessage = "Booom!"
+      Source.failed(new Exception(remoteFailureMessage))
+        .to(remoteSink)
+        .run()
+
+      val f = p.expectMsgType[akka.actor.Status.Failure]
+      f.cause.getMessage should include(s"Remote stream (")
+      // actor name ere, for easier identification
+      f.cause.getMessage should include(s"failed, reason: $remoteFailureMessage")
+    }
+
+    "receive hundreds of elements via remoting" in {
+      remoteActor ! "receive"
+      val remoteSink: SinkRef[String] = expectMsgType[SinkRef[String]]
+
+      val msgs = (1 to 100).toList.map(i ⇒ s"payload-$i")
+
+      Source(msgs)
+        .runWith(remoteSink)
+
+      msgs.foreach(t ⇒ p.expectMsg(t))
+      p.expectMsg("<COMPLETE>")
+    }
+
```

Comment:
perhaps some test that verifies backpressure, I think that is kind of "lost" in tests using `Sink.actorRef` ?
perhaps use stream test probes for that test

Reply:
Oh, excellent point, definitely a test to add

Reply:
Added, fixed a thing I missed in there